### PR TITLE
Adding ISIS3 Style Kernel Selection 

### PR DESCRIPTION
**Pull request by Kelvinrr**
Tested by hand by comparing Kernel groups of Spiceinited cubes, pytests incoming.

### CODE REVIEW

File: ale/util.py [hunk 1]
```diff
@@ -506,3 +501,303 @@ def query_kernel_pool(matchstr="*"):
...
+
+    Returns : PVLModule
+              object representing the parsed PVL
+
+
+    """
+    def JBFKeywordParser(lines):
+        keyword = lines[0].split("=")[0]
+        value = lines[0].split("=")[1].strip()
+        value = lines[0].split("=")[1]+"".join(l.strip() for l in lines[1:])
```
Comment (jessemapel) [inline ale/util.py]: This looks like duplicate lines

File: ale/util.py [hunk 2]
```diff
@@ -506,3 +501,303 @@ def query_kernel_pool(matchstr="*"):
...
+            metadata.append([i, "keyword"])
+        elif "end_group" in l.lower() or "endgroup" in l.lower():
+            metadata.append([i, "group_end"])
+        elif "end_object" in l.lower() or "endobject" in l.lower():
+            metadata.append([i, "object_end"])
+
+    imeta = 0
+    while imeta < len(metadata):
+        m = metadata[imeta]
+        element_start_line, element_type = m
```
Comment (jessemapel) [inline ale/util.py]: Why not unpack in a single line?

File: ale/util.py [hunk 3]
```diff
@@ -506,3 +501,303 @@ def query_kernel_pool(matchstr="*"):
...
+
+    lines : str
+          string body of PVL file.
+
+    Returns : PVLModule
+              object representing the parsed PVL
+
+
+    """
+    def JBFKeywordParser(lines):
```
Comment (jessemapel) [inline ale/util.py]: This looks like it only works with a single line. Why pass in a list?

File: ale/util.py [hunk 4]
```diff
@@ -506,3 +501,303 @@ def query_kernel_pool(matchstr="*"):
...
+    utc_stop_time = labelobj["IsisCube"]["Instrument"]["StopTime"]
+
+    run_time = None
+    dependencies = None
+    kernels = []
+    typ = None
+    types = []
+
+    # flag is set when a kernel is found matching the start time but not stop time,
+    # this may
```
Comment (jessemapel) [inline ale/util.py]: Partially complete comment?

File: ale/util.py [hunk 5]
```diff
@@ -506,3 +501,303 @@ def query_kernel_pool(matchstr="*"):
...
+
+        if any(matches if matches else [True]):
+            for i,f in enumerate(files):
+                if isinstance(f, tuple):
+                    f = os.path.join(*f)
+
+                full_path = os.path.join(isis3_data, f).replace("$", "")
+                if '?' in full_path:
+                    full_path = sorted(glob(full_path))[-1]
+                files[i] = full_path
```
Comment (jessemapel) [inline ale/util.py]: There are some files that use more complex Isis::FileName matching. For example, here's what the Cassini PCK selection looks like:

```
  Group = Selection
    Time = ("1999 AUG 19 02:00:00.000 TDB", "2001 MAR 24 00:00:00.000 TDB")
    File = (base, kernels/pck/pck?????.tpc)
    File = "$cassini/kernels/pck/cpck{ddMMMyyyy}_jupiter.tpc"
  End_Group
```

File: ale/util.py [hunk 6]
```diff
@@ -506,3 +501,303 @@ def query_kernel_pool(matchstr="*"):
...
+                    full_match = True
+                    kernels = files
+                    types = [selection.get("Type", None)]
+
+                elif any(have_start_match):
+                    kernels = files
+                    types.append(selection.get("Type", None))
+                    looking_for_end = True
+                elif any(have_stop_match):
+                    if partial_match:
```
Comment (jessemapel) [inline ale/util.py]: partial_match never seems to be set.

File: ale/util.py (hunk 3 above)
Comment (Kelvinrr) [reply ale/util.py]: Keywords can make up multiple lines. Arrays and long wrapping keywords are two common examples.

**Comment by jessemapel** [comment]
@Kelvinrr Feel free to merge this once you're happy with the testing.

**Comment by Kelvinrr** [comment]
I think it's good enough for now until I can revisit.

**Pull request merged**